src/test/java/com/emc/object/s3/S3JerseyClientTest.java

```diff
@@ -2942,4 +2943,101 @@ protected void assertAclEquals(AccessControlList acl1, AccessControlList acl2) {
         Assert.assertEquals(acl1.getOwner(), acl2.getOwner());
         Assert.assertEquals(acl1.getGrants(), acl2.getGrants());
     }
+
+    @Test
+    public void testGetObjectTagging() {
+        // set up env
+        String bucketName = getTestBucket(), okey = "test-object-tagging";
+        client.setBucketVersioning(bucketName, new VersioningConfiguration().withStatus(VersioningConfiguration.Status.Enabled));
+        // have the versions with different tagging settings.
+        PutObjectRequest putObjectRequest1 = new PutObjectRequest(bucketName, okey, "Hello Version !");
+        client.putObject(putObjectRequest1);
+        String versionId1 = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+        PutObjectRequest putObjectRequest2 = new PutObjectRequest(bucketName, okey, "Hello Version !")
+                .withTagging(new HashMap<String, String>(){{put("k1", "v1");}});
+        client.putObject(putObjectRequest2);
+        String versionId2 = client.listVersions(bucketName, null).getVersions().get(0).getVersionId();
+        // should be able to read object tags associated with an object
+        Tagging t = client.getObjectTagging(new GetObjectTaggingRequest(bucketName, okey).withVersionId(versionId1));
+        Assert.assertNull(t.getTagSet());
+        t = client.getObjectTagging(new GetObjectTaggingRequest(bucketName, okey).withVersionId(versionId2));
+        Assert.assertEquals("k1v1", t.getTagSet().getTags().get(0).getKey() + t.getTagSet().getTags().get(0).getValue());
+        // GET the Tag of a non-existent object
+        try {
+            t = client.getObjectTagging(new GetObjectTaggingRequest(bucketName, "object-key-not-exist"));
+        } catch (S3Exception e) {
+            Assert.assertEquals(404, e.getHttpCode());
+            Assert.assertEquals("NoSuchKey", e.getErrorCode());
+        }
+        // GET the tag of an object where the tag was previously deleted.
+        client.deleteObjectTagging(new DeleteObjectTaggingRequest(bucketName, okey).withVersionId(versionId2));
+        t = client.getObjectTagging(new GetObjectTaggingRequest(bucketName, okey).withVersionId(versionId2));
+        Assert.assertNull(t.getTagSet());
+
+    }
+
+    @Test
+    public void testPutObjectTagging() {
+        // set up env
+        String bucketName = getTestBucket(), okey = "test-object-tagging";
+        List<Tag> tag = new ArrayList<Tag>(){{add(new Tag().withKey("k0").withValue("v0"));}}; // a new [single] tag
+        List<Tag> tags = new ArrayList<Tag>(){{ // multiple tags[10]
+            add(new Tag().withKey("k1").withValue("v1"));
+            add(new Tag().withKey("k2").withValue("v2"));
+            add(new Tag().withKey("k3").withValue("v3"));
+            add(new Tag().withKey("k4").withValue("v4"));
```

Comment:
It looks like the number of tags(10 here) won't make much difference, maybe should verify some special character?